packages/cli/src/cmds/validator/keymanager/persistedKeys.ts

```diff
@@ -114,6 +118,65 @@ export class PersistedKeysBackend implements IPersistedKeysBackend {
     return keystoreDefinitions;
   }
 
+  async readKeyStoresInBatches(files: string[], path: string, batchSize: number): Promise<LocalKeystoreDefinition[]> {
+    const numBatches = Math.ceil(files.length / batchSize);
+
+    const batches = Array.from({length: numBatches}, (_, i) => {
+      const start = i * batchSize;
+      const end = start + batchSize;
+      return files.slice(start, end);
+    });
+
+    const allKeystoreDefinitions = await Promise.all(
+      batches
+        .map((batch) => {
+          return new Promise<LocalKeystoreDefinition[]>((resolve, reject) => {
+            return this.readKeystoreByBatch(path, batch, resolve, reject);
+          });
+        })
+        .flat()
+    );
+
+    return allKeystoreDefinitions.flat();
+  }
+
+  readKeystoreByBatch(
+    path: string,
+    batch: string[],
+    resolve: (value: LocalKeystoreDefinition[]) => void,
+    reject: (err: Error) => void
```

Comment:
I'm very confused by this pattern, why are the Promise callbacks being passed as arguments?

Reply:
The idea was break the file down into batches, and then read from the file in batches of 5-10.  I think using [worker threads](https://nodejs.org/api/worker_threads.html) may make a better alternative though.

Reply:
> I think using [worker threads](https://nodejs.org/api/worker_threads.html) may make a better alternative though

isn't most of work of BLS Keystore `decrypt` executed in a separate thread already?

Reply:
Yes but the process of reading the files itself is not multithreaded, nor is the decryption of these multiple keys, which is what I assumed was the rationale behind 

> Instead of decrypting 100 keystores at once, decrypting 5-10 is fine, which provides a significant speed up. A good number would be to decrypt thread count at once

and perhaps the bottleneck for situations like this https://twitter.com/Stefa2kEth/status/1628407293325807623?s=20

Reply:
> nor is the decryption of these multiple keys

would have to dig into how bls-keystore implements this but I just assumed this was not executed in the main thread

> and perhaps the bottleneck for situations like this

the main issue is that we decrypt just one at a time, and running all in sequence caused a OOM error

Would be curious if running this in a worker actually makes a noticeable difference

Reply:
would be interesting to see what acutally takes up the time in `decryptKeystoreDefinitions` https://github.com/ChainSafe/lodestar/blob/df1ef97e9050a588b74f21bfae314d675caea288/packages/cli/src/cmds/validator/keymanager/persistedKeys.ts#L257

I just assumed as of now that `keystore.decrypt` takes the vast majority of the time and it should not be executed in the main thread, I would be really surprised if it was
 https://github.com/ChainSafe/lodestar/blob/df1ef97e9050a588b74f21bfae314d675caea288/packages/cli/src/cmds/validator/keymanager/persistedKeys.ts#L304